lib/semantics/resolve-names.cc

```diff
@@ -1037,11 +1037,271 @@ class ConstructVisitor : public DeclarationVisitor {
   void PopAssociation();
 };
 
+static const parser::Name *GetDesignatorNameIf(
+    const parser::Designator &designator) {
+  const auto *dataRef{std::get_if<parser::DataRef>(&designator.u)};
+  return dataRef ? std::get_if<parser::Name>(&dataRef->u) : nullptr;
+}
+
+static constexpr Symbol::Flags dataSharingAttributeFlags{
+    Symbol::Flag::OmpShared, Symbol::Flag::OmpPrivate,
+    Symbol::Flag::OmpFirstPrivate, Symbol::Flag::OmpLastPrivate,
+    Symbol::Flag::OmpReduction, Symbol::Flag::OmpLinear};
+
+static constexpr Symbol::Flags ompFlagsRequireNewSymbol{
+    Symbol::Flag::OmpPrivate, Symbol::Flag::OmpLinear,
+    Symbol::Flag::OmpFirstPrivate, Symbol::Flag::OmpLastPrivate,
+    Symbol::Flag::OmpReduction};
+
+static constexpr Symbol::Flags ompFlagsRequireMark{
+    Symbol::Flag::OmpThreadprivate};
+
+// Resolve OpenMP construct entities and statement (TODO) entities
+class OmpVisitor : public virtual DeclarationVisitor {
+public:
+  bool Pre(const parser::OpenMPBlockConstruct &) {
+    PushScope(Scope::Kind::Block, nullptr);
+    return true;
+  }
+  void Post(const parser::OpenMPBlockConstruct &) { PopScope(); }
+  bool Pre(const parser::OmpBeginBlockDirective &) {
+    clear_dataSharingAttributeObject();
+    return true;
+  }
+  void Post(const parser::OmpBeginBlockDirective &) {
+    clear_dataSharingAttributeObject();
+  }
+
+  bool Pre(const parser::OpenMPLoopConstruct &) {
+    PushScope(Scope::Kind::Block, nullptr);
+    return true;
+  }
+  void Post(const parser::OpenMPLoopConstruct &) { PopScope(); }
+  bool Pre(const parser::OmpBeginLoopDirective &) {
+    clear_dataSharingAttributeObject();
+    return true;
+  }
+  void Post(const parser::OmpBeginLoopDirective &) {
+    clear_dataSharingAttributeObject();
+  }
+
+  bool Pre(const parser::OpenMPThreadprivate &x) {
+    PushScope(Scope::Kind::Block, nullptr);
+    const auto &list{std::get<parser::OmpObjectList>(x.t)};
+    ResolveOmpObjectList(list, Symbol::Flag::OmpThreadprivate);
+    PopScope();
+    return false;
+  }
+
+  bool Pre(const parser::OmpClause::Shared &x) {
+    ResolveOmpObjectList(x.v, Symbol::Flag::OmpShared);
+    return false;
+  }
+  bool Pre(const parser::OmpClause::Private &x) {
+    ResolveOmpObjectList(x.v, Symbol::Flag::OmpPrivate);
+    return false;
+  }
+  bool Pre(const parser::OmpClause::Firstprivate &x) {
+    ResolveOmpObjectList(x.v, Symbol::Flag::OmpFirstPrivate);
+    return false;
+  }
+  bool Pre(const parser::OmpClause::Lastprivate &x) {
+    ResolveOmpObjectList(x.v, Symbol::Flag::OmpLastPrivate);
+    return false;
+  }
+
+  void add_dataSharingAttributeObject(const Symbol *object) {
+    dataSharingAttributeObjects_.insert(object);
+  }
+  void clear_dataSharingAttributeObject() {
+    dataSharingAttributeObjects_.clear();
+  }
+  bool find_dataSharingAttributeObject(const Symbol *);
```

Comment:
This isn't returning something that is found, so a name that starts with `Is` or `Has` is probably more appropriate. It might also make it clearer what it means for a symbol to be in `dataSharingAttributeObjects_`.

Reply:
Also, should the argument be `const Symbol &`?